### PR TITLE
Fixes #1503 : Add RatioInput Module

### LINKED ISSUE #1503: Add RatioInput Module
Introduce the following things in order to add Ratio input Interaction support

- [ ] Annotation class for Ratio Input Rules classfiers

- [ ] isEquals Classifier

- [ ] Declaring RatioInput in application component

- [ ] Registering Interaction inside InteractionObjectTypeExtractorRepository

### PR DESCRIPTION
**Pull request by aggarwalpulkit596**
## Explanation

Fixes part of #1503 

- Add Equals Classifier

- Add RatioInputModule 

This PR will add the ratio interaction module which is required for any interaction to get started with and each module for an interaction consists of different rules/classifiers which helps in verifying the user's input through different criteria.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioExpressionInputModule.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import dagger.multibindings.StringKey
+import org.oppia.domain.classify.RuleClassifier
+import org.oppia.domain.classify.rules.RatioExpressionInputRules
+
+/** Module that binds rule classifiers corresponding to the ratio input interaction. */
+@Module
+class RatioExpressionInputModule {
+  @Provides
+  @IntoMap
+  @StringKey("Equals")
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioExpressionInputModule.kt]: What abouy equivalence? Is that being implemented in a later PR?

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioInputEqualsRuleClassifierProvider.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import org.oppia.domain.classify.RuleClassifier
+import org.oppia.domain.classify.rules.GenericRuleClassifier
+import org.oppia.domain.classify.rules.RuleClassifierProvider
+import org.oppia.domain.util.normalizeWhitespace
+import javax.inject.Inject
+
+/**
+ * Provider for a classifier that determines whether two object are equal per the ratio input
+ * interaction.
+ *
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioInputEqualsRuleClassifierProvider.kt]: Remove extra newline in KDoc.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @aggarwalpulkit596. Please fix the branching for this PR—it seems to include changes from your other PR. 

Also, will do another full pass after the interaction object is updated.

**Comment by vinitamurthi** [comment]
I am taking myself out of reviewers for this one since I think both @rt4914 and @BenHenning have more context on this as me

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioExpressionInputModule.kt (hunk 1 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioExpressionInputModule.kt]: Every Rule is a seperate PR

File: domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  @Test
+  fun testAnswer_testRatio_ratio1_bothValuesMatch() {
+    val inputs = mapOf("x" to createRatio(ITEM_RATIO_1))
+
+    val matches =
+      equalsClassifierProvider.matches(answer = createRatio(ITEM_RATIO_1), inputs = inputs)
+
+    Truth.assertThat(matches).isTrue()
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt]: Directly import ``assertThat``

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @aggarwalpulkit596. There are a few more comments to address yet.

File: domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+import kotlin.test.fail
+
+/** Tests for [RatioInputEqualsRuleClassifierProvider]. */
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(manifest = Config.NONE)
+class RatioInputEqualsRuleClassifierProviderTest {
+  private val NON_NEGATIVE_VALUE_0 = createNonNegativeInt(value = 0)
+  private val ITEM_RATIO_1 = listOf(1, 2, 3)
+  private val ITEM_RATIO_2 = listOf(2, 4, 6)
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt]: Add tests for ratios of different lengths, too.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Added another comment. Also, please fix the PR title & clarify in the description what this PR is actually introducing. I don't think this is image click input, and you should explain what the changes are & how they're part of the broader project. Linking to specific sections of the design doc is also good.

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioInputEqualsRuleClassifierProvider.kt (hunk 2 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioInputEqualsRuleClassifierProvider.kt]: done

File: domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt (hunk 4 above)
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt]: done

File: domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt (hunk 3 above)
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt]: done

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioInputEqualsRuleClassifierProvider.kt (hunk 2 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/ratioExpressionInput/RatioInputEqualsRuleClassifierProvider.kt]: done

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProvider.kt [hunk 5]
```diff
@@ -1,17 +1,15 @@
-package org.oppia.domain.classify.rules.ratioExpressionInput
+package org.oppia.domain.classify.rules.ratioInput
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProvider.kt]: Ditto here and in your other PRs. Should be fully lowercase per https://developer.android.com/kotlin/style-guide#package_names.

File: domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt [hunk 6]
```diff
@@ -45,17 +45,27 @@ class RatioInputEqualsRuleClassifierProviderTest {
...
 
     val matches =
       equalsClassifierProvider.matches(answer = createRatio(ITEM_RATIO_1), inputs = inputs)
 
-    Truth.assertThat(matches).isFalse()
+    assertThat(matches).isFalse()
+  }
+
+  @Test
+  fun testAnswer_testRatio_ratio1_ratio3_bothValuesDoNoMatch() {
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt]: These test names aren’t super clear. How do we verify the test is verifying the correct thing from its name? It’s not that clear what the differences between these different ratios are in order to validate. Please clarify the actual conditions being tested in each test.

**Review by BenHenning — CHANGES_REQUESTED** [review]

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProvider.kt (hunk 5 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProvider.kt]: done

File: domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt (hunk 6 above)
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/ratioInput/RatioInputEqualsRuleClassifierProviderTest.kt]: updated the test names let me know if they aren't clear enough

File: model/src/main/proto/interaction_object.proto [hunk 7]
```diff
@@ -31,8 +31,9 @@ message StringList {
   repeated string html = 1;
 }
 
-// Structure containing a ratio object for eg - [1,2,3].
+// Structure containing a ratio object for eg - [1,2,3] for 1:2:3.
 message RatioExpression {
+  // List of components in a ratio.
   // It's also expected that list should have more than 1 elements with
   // 0 and 1 elements are invalid states.
```
Comment (BenHenning) [inline model/src/main/proto/interaction_object.proto]:
```suggestion
  // List of components in a ratio. It's expected that list should have more than
  // 1 element.
```

**Review by BenHenning — APPROVED** [review]
Left one comment that I think came from the proto PR, but otherwise this LGTM. Feel free to resolve the remaining comment thread & merge after applying the fix (though note that it might need reformatting).

**Pull request merged**